From 6189b637f9ec176eb28f32ecf0336792a3a933f2 Mon Sep 17 00:00:00 2001
From: Jeff Moyer <jmoyer@redhat.com>
Date: Thu, 28 Aug 2014 19:16:09 -0400
Subject: [virt] virtio_blk: introduce feature of VIRTIO_BLK_F_MQ

Message-id: <1409253371-23161-20-git-send-email-jmoyer@redhat.com>
Patchwork-id: 89852
O-Subject: [RHEL7 PATCH 19/21] include/uapi/linux/virtio_blk.h: introduce feature of VIRTIO_BLK_F_MQ
Bugzilla: 1105698
RH-Acked-by: Fam Zheng <famz@redhat.com>

This is a backport of the following upstream commit.
Related to bug 1105698.

  commit cb553215d5d277d4838d7d6b7722e964bcf5ca1f
  Author: Ming Lei <ming.lei@canonical.com>
  Date:   Thu Jun 26 17:41:47 2014 +0800

    include/uapi/linux/virtio_blk.h: introduce feature of VIRTIO_BLK_F_MQ

    Current virtio-blk spec only supports one virtual queue for transfering
    data between VM and host, and inside VM all kinds of operations on
    the virtual queue needs to hold one lock, so cause below problems:

     - bad scalability
     - bad throughput

    This patch requests to introduce feature of VIRTIO_BLK_F_MQ
    so that more than one virtual queues can be used to virtio-blk
    device, then above problems can be solved or eased.

    Signed-off-by: Ming Lei <ming.lei@canonical.com>
    Acked-by: Michael S. Tsirkin <mst@redhat.com>
    Signed-off-by: Jens Axboe <axboe@fb.com>

Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/uapi/linux/virtio_blk.h b/include/uapi/linux/virtio_blk.h
index 6d8e61c..9ad67b2 100644
--- a/include/uapi/linux/virtio_blk.h
+++ b/include/uapi/linux/virtio_blk.h
@@ -40,6 +40,7 @@
 #define VIRTIO_BLK_F_WCE 9 /* Writeback mode enabled after reset */
 #define VIRTIO_BLK_F_TOPOLOGY 10 /* Topology information is available */
 #define VIRTIO_BLK_F_CONFIG_WCE 11 /* Writeback mode available in config */
+#define VIRTIO_BLK_F_MQ  12 /* support more than one vq */
 
 #ifndef __KERNEL__
 /* Old (deprecated) name for VIRTIO_BLK_F_WCE. */
@@ -77,6 +78,10 @@ struct virtio_blk_config {
 
  /* writeback mode (if VIRTIO_BLK_F_CONFIG_WCE) */
  __u8 wce;
+ __u8 unused;
+
+ /* number of vqs, only available when VIRTIO_BLK_F_MQ is set */
+ __u16 num_queues;
 } __attribute__((packed));
 
 /*
-- 
1.7.1